provisionersdk/session.go

```diff
@@ -0,0 +1,298 @@
+package provisionersdk
+
+import (
+	"archive/tar"
+	"bytes"
+	"cdr.dev/slog"
+	"context"
+	"fmt"
+	"golang.org/x/xerrors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/coder/coder/v2/provisionersdk/proto"
+)
+
+// ReadmeFile is the location we look for to extract documentation from template
+// versions.
+const ReadmeFile = "README.md"
+
+// protoServer is a wrapper that translates the dRPC protocol into a Session with method calls into the Server.
+type protoServer struct {
+	server Server
+	opts   ServeOptions
+}
+
+func (p *protoServer) Session(stream proto.DRPCProvisioner_SessionStream) error {
+	s := &Session{
+		Logger: p.opts.Logger,
+		stream: stream,
+		server: p.server,
+	}
+	sessDir := fmt.Sprintf("Session%p", s)
+	s.WorkDirectory = filepath.Join(p.opts.WorkDirectory, sessDir)
+	err := os.MkdirAll(s.WorkDirectory, 0o700)
+	if err != nil {
+		return xerrors.Errorf("create work directory %q: %w", s.WorkDirectory, err)
+	}
+	defer func() {
+		// Cleanup the work directory after execution.
+		for attempt := 0; attempt < 5; attempt++ {
+			err := os.RemoveAll(s.WorkDirectory)
```

Comment:
Would it be useful to add a config option to keep the directory to template/coder debugging?

Reply:
Maybe?; but beyond the scope of the refactor.  This is what the daemon `runner.go` does today.